sql/2024/fonts/common.sql

```diff
@@ -0,0 +1,123 @@
+-- Normalize a family name. Used in FAMILY_INNER.
+CREATE TEMPORARY FUNCTION FAMILY_INNER_INNER(name STRING) AS (
+  CASE
+    WHEN REGEXP_CONTAINS(name, r'(?i)font\s?awesome') THEN 'Font Awesome'
+    ELSE IF(LENGTH(TRIM(name)) < 3, NULL, NULLIF(TRIM(name), ''))
+  END
+);
+
+-- Normalize a family name. Used in FAMILY.
+CREATE TEMPORARY FUNCTION FAMILY_INNER(name STRING) AS (
+  FAMILY_INNER_INNER(
+    REGEXP_REPLACE(
+      name,
+      r'(?i)([\s-]?(black|bold|book|cond(ensed)?|demi|ex(tra)?|heavy|italic|light|medium|narrow|regular|semi|thin|ultra|wide|\d00|\d+pt))+$',
+      ''
+    )
+  )
+);
+
+-- Extract the family name from a payload.
+CREATE TEMPORARY FUNCTION FAMILY(payload STRING) AS (
+  FAMILY_INNER(
+    COALESCE(
+      JSON_EXTRACT_SCALAR(payload, '$._font_details.names[16]'),
+      JSON_EXTRACT_SCALAR(payload, '$._font_details.names[1]')
+    )
+  )
+);
+
+-- Extract the file format from an extension and a MIME type.
+CREATE TEMPORARY FUNCTION FILE_FORMAT(extension STRING, type STRING) AS (
+  LOWER(IFNULL(REGEXP_EXTRACT(type, '/(?:x-)?(?:font-)?(.*)'), extension))
+);
+
+-- Extract the foundry name from a payload.
+CREATE TEMPORARY FUNCTION FOUNDRY(payload STRING) AS (
+  NULLIF(
+    TRIM(REGEXP_REPLACE(
+      JSON_EXTRACT_SCALAR(payload, '$._font_details.OS2.achVendID'),
+      r'[[:cntrl:]]+',
+      ''
+    )),
+    ''
+  )
+);
+
+-- Infer scripts from codepoints. Used in SCRIPTS.
+CREATE TEMPORARY FUNCTION SCRIPTS_INNER(codepoints ARRAY<STRING>)
+RETURNS ARRAY<STRING>
+LANGUAGE js
+OPTIONS (library = ["gs://httparchive/lib/text-utils.js"])
+AS r"""
+if (codepoints && codepoints.length) {
+  return detectWritingScript(codepoints.map((character) => parseInt(character, 10)), 0.05);
+} else {
+  return [];
+}
+""";
+
+-- Infer scripts from a payload.
+CREATE TEMPORARY FUNCTION SCRIPTS(payload STRING) AS (
```

Comment:
I worry about people confusing scripts with JavaScript (I think I did earlier in the chat for one!). Should we call this something else? WRITING_SCRIPT? FONT_SCRIPT? CODEPOINTS?

Reply:
We have only one query dealing with JavaScript scripts (`performance/scripts_font_face.sql`). I think there will be little confusion what _script_ refers to when browsing the queries for the Fonts chapter. I am also assuming that in the article, one will be careful to explain.